### PR TITLE
Add cryptographic verification to Release

### PR DESCRIPTION
**Pull request by mdellweg**

### CODE REVIEW

File: plugins/pulp_deb/plugins/importers/sync.py [hunk 1]
```diff
@@ -129,6 +135,20 @@ def process_main(self, item=None):
         architectures = self.parent.architectures
         dl_reqs = []
         for release in releases:
+            if not os.path.isfile(self.parent.release_files[release]):
```
Comment (mibanescu) [inline plugins/pulp_deb/plugins/importers/sync.py]: Can you separate this block into its own function, so it's easier to test?

File: pulp-dev.py [hunk 2]
```diff
@@ -40,7 +40,7 @@
      '/etc/pulp/server/plugins.conf.d/deb_importer.json'),
 )
 
-PIP_INSTALLABLE = set(['debpkgr'])
+PIP_INSTALLABLE = set(['debpkgr', 'gnupg'])
```
Comment (mibanescu) [inline pulp-dev.py]: I think the spec file needs to also be edited to pull in python2-gnupg (an EPEL package, as far as I can tell). I am checking with pulp's release engineering team to see if they have to build it into the pulp repos.

File: plugins/pulp_deb/plugins/importers/sync.py (hunk 1 above)
Comment (mdellweg) [reply plugins/pulp_deb/plugins/importers/sync.py]: you know, i am a bit lazy on the tests until i know, how the structure of the feature is going to settle. Tests will follow.

File: pulp-dev.py (hunk 2 above)
Comment (mdellweg) [reply pulp-dev.py]: Will do.

**Comment by mdellweg** [comment]
Thanks for the detailed review.
For the signature checking policy i was inspired by apt's way, where you add a bunch of keys with 'apt-key' and i really does not matter who signed what.
Are you suggesting to specify the complete key fingerprint in the 'verify-release-with', or even better a list of viable signing keys fprs?

**Comment by mibanescu** [comment]
Yes, I was thinking a list of viable signing key fingerprints. While it may not be exactly what apt does, I think there is a reason. pulp is inherently a multi-user system. More than one group may be using it. They may manage (and mirror) different sets of apt repositories, and having shared keyrings does not seem appropriate.

**Comment by mdellweg** [comment]
Fair enough! And if we only accept whole fingerprints, it is as secure as it can get. ACK.

**Comment by mdellweg** [comment]
@mibanescu Is this, what you thought about?

**Comment by mibanescu** [comment]
Before I look at the PR, I wanted to share what the Pulp developers pointed me to: https://github.com/pulp/pulp_rpm/commit/01b21befdd719884ca0044253fbbe60c2f65f68c - the signature checking implemented in yum.

File: pulp-deb.spec [hunk 3]
```diff
@@ -91,6 +91,7 @@ Requires: python-%{name}-common = %{version}-%{release}
 Requires: pulp-server >= %{pulp_version}
 Requires: python-debian
 Requires: python-debpkgr >= 1.0.0
+Requires: python-gnupg
```
Comment (mibanescu) [inline pulp-deb.spec]: I think it's python2-gnupg in EPEL repos.

Comment (mdellweg) [reply pulp-deb.spec]: Yes, i keep mixing this with the debian package names...

**Comment by mdellweg** [comment]
@mibanescu I think, that should be it (with a test ^_^)

File: plugins/pulp_deb/plugins/importers/sync.py [hunk 4]
```diff
@@ -123,12 +128,40 @@ class ParseReleaseStep(publish_step.PluginStep):
     def __init__(self, *args, **kwargs):
         super(ParseReleaseStep, self).__init__(*args, **kwargs)
         self.description = _('Parse Release Files')
+
+    def verify_release(self, release):
+        # check if Release file exists
+        if not os.path.isfile(self.parent.release_files[release]):
+            raise Exception("Release file not found. Check the feed option.")
+        # check signature
+        if self.get_config().get_boolean(constants.CONFIG_REQUIRE_SIGNATURE, False):
```
Comment (mibanescu) [inline plugins/pulp_deb/plugins/importers/sync.py]: As a matter of personal preference, I would test the negative and return. That way the rest of the code is one level of indentation less, and seems easier to follow. But everything else looks good, and if you prefer not to change that part, let me know, and I will merge it.

**Comment by mdellweg** [comment]
`Requires: python2-gnupg`
Has been lost while rebasing, due to the spec-file being deleted.

**Pull request merged**